src/cat.jl

```diff
@@ -0,0 +1,135 @@
+
+"""
+    ConcatDiskArray <: AbstractDiskArray
+
+Joins multiple AbstractArrays or AbstractDiskArrays in lazy concatination.
+"""
+struct ConcatDiskArray{T,N,P} <: AbstractDiskArray{T,N}
+    parents::P
+    startinds::NTuple{N,Vector{Int}}
+    size::NTuple{N,Int}
+end
+function ConcatDiskArray(arrays::AbstractArray{<:AbstractArray{T,N},M}) where {T,N,M}
+    function othersize(x, id)
+        (x[1:id-1]..., x[id+1:end]...)
+    end
+    if N > M
+        newshape = (size(arrays)..., ntuple(_ -> 1, N - M)...)
+        arrays1 = reshape(arrays, newshape)
+        D = N
+    elseif N < M
+        arrays1 = map(arrays) do a
+            newshape = (size(a)..., ntuple(_ -> 1, M - N)...)
+            reshape(a, newshape)
+        end
+        D = M
+    else
+        arrays1 = arrays
+        D = M
+    end
+    arraysizes = map(size, arrays1)
+    si = ntuple(D) do id
+        a = reduce(arraysizes; dims=id, init=ntuple(zero, D)) do i, j
+            if all(iszero, i)
+                j
+            elseif othersize(i, id) == othersize(j, id)
+                j
+            else
+                error("Dimension sizes don't match")
+            end
+        end
+        I = ntuple(D) do i
+            i == id ? Colon() : 1
+        end
+        ari = map(i -> i[id], arraysizes[I...])
+        sl = sum(ari)
+        r = cumsum(ari)
+        pop!(pushfirst!(r, 0))
+        r .+ 1, sl
+    end
+
+    startinds = map(first, si)
+    sizes = map(last, si)
+
+    return ConcatDiskArray{T,D,typeof(arrays1)}(arrays1, startinds, sizes)
+end
+function ConcatDiskArray(arrays::AbstractArray)
+    # Validate array eltype and dimensionality
+    all(t -> t == eltype(first(arrays)), eltypes) || error("Arrays don't have the same element type")
+    all(s -> length(s) == ndims(first(arrays)), sizes) || error("Arrays don't have the same dimensions")
+    error("Should not be reached")
+end
+
+Base.size(a::ConcatDiskArray) = a.size
+
+function readblock!(a::ConcatDiskArray, aout, inds::AbstractUnitRange...)
+    # Find affected blocks and indices in blocks
+    blockinds = map(inds, a.startinds, size(a.parents)) do i, si, s
+        bi1 = max(searchsortedlast(si, first(i)), 1)
+        bi2 = min(searchsortedfirst(si, last(i) + 1) - 1, s)
+        bi1:bi2
+    end
+    map(CartesianIndices(blockinds)) do cI
+        myar = a.parents[cI]
+        mysize = size(myar)
+        array_range = map(cI.I, a.startinds, mysize, inds) do ii, si, ms, indstoread
+            max(first(indstoread) - si[ii] + 1, 1):min(last(indstoread) - si[ii] + 1, ms)
+        end
+        outer_range = map(cI.I, a.startinds, array_range, inds) do ii, si, ar, indstoread
+            (first(ar)+si[ii]-first(indstoread)):(last(ar)+si[ii]-first(indstoread))
+        end
+        aout[outer_range...] = a.parents[cI][array_range...]
+    end
+end
+
+function writeblock!(a::ConcatDiskArray, aout, inds::AbstractUnitRange...)
```

Comment:
Do you think we need this? It would be more or less copy/paste of `readblock!` but with LHS and RHS reversed in line 81. So in principle better to make a generic IO function but we can defer this to later as well. 

The main use case would be if someone wanted to write results into a pre-generated annual array of netcdf files or similar.

Reply:
That sounds useful. I'll move the `readblock!` code to a function and call it in botg